Removing probability field from answers in favor of score field

julian-risch opened the pull request:
**Proposed changes**:
This PR drops the field `probability` from answers and documents in favor of the field `score`. The default behavior is now that the field `score` contains the value that was previously stored in probability. However, the FARMReader also allows switching back to the old scores by setting `use_confidence_scores` to False (default is True).
closes #1220 

Main changes:
* no_answer prediction now uses score ranging between 0 and 1: https://github.com/deepset-ai/haystack/blob/dda0e857f1ea3d7272867e28ce7e900647231176/haystack/reader/base.py#L45
* FARMReader allows switching between new scores ranging between 0 and 1 (confidence scores) and old scores: https://github.com/deepset-ai/haystack/blob/dda0e857f1ea3d7272867e28ce7e900647231176/haystack/reader/farm.py#L55
* FARMReader extracts new or old scores from QAPreds depending on this switch: https://github.com/deepset-ai/haystack/blob/dda0e857f1ea3d7272867e28ce7e900647231176/haystack/reader/farm.py#L577
* I relaxed an assertion in one of the test cases. Previously, it checked that the correct answer is the top prediction. It had the best score but only the second best probability so it worked in the past. Now, I changed the test so that it checks whether the correct predictions is within the top 2 predictions as the order of the ranked predictions slightly changed.

@tholor I came across some points in the code where a plain dictionary could be replaced with an Answer object:
https://github.com/deepset-ai/haystack/blob/dda0e857f1ea3d7272867e28ce7e900647231176/haystack/reader/transformers.py#L134
https://github.com/deepset-ai/haystack/blob/dda0e857f1ea3d7272867e28ce7e900647231176/haystack/generator/transformers.py#L278
https://github.com/deepset-ai/haystack/blob/dda0e857f1ea3d7272867e28ce7e900647231176/haystack/reader/base.py#L44
https://github.com/deepset-ai/haystack/blob/dda0e857f1ea3d7272867e28ce7e900647231176/haystack/reader/farm.py#L576
https://github.com/deepset-ai/haystack/blob/dda0e857f1ea3d7272867e28ce7e900647231176/haystack/pipeline.py#L602

**Status (please check what you already did)**:
- [x] First draft (up for discussions & feedback)
- [ ] Final code
- [x] Added tests -> I added one test case that checks the behavior of the newly introduced switch `use_confidence_scores` in FARMReader. If `use_confidence_scores` is set to False, the old scores are used, which can be larger than 1.0
- [x] Updated documentation -> I removed mentions of "probability" from docs

tholor commented on haystack/document_store/elasticsearch.py:
```diff
@@ -872,7 +871,6 @@ def _convert_es_hit_to_document(
             text=hit["_source"].get(self.text_field),
             meta=meta_data,
             score=score,
-            probability=probability,
```
We can simplify the if-block above as probability is not needed anymore. Probably we should apply the scaling part to the `score` now to be again in range 0-1

julian-risch replied:
Agreed. I will simplify the if-block and apply the scaling to the `score` variable.

tholor reviewed (commented):
Looks all good to me. Two questions: 
- I assume Summarizer, Re-Ranker and classifier did not use any probabilities before?
- Should we scale `score` always between 0-1 now? I think this would simplify communication to users a lot. I assume this scaling is currently not the case for Elasticsearch (see comment).

tholor commented on docs/_src/api/api/reader.md:
```diff
@@ -158,7 +158,7 @@ Saves the Reader model so that it can be reused at a later point in time.
 Use loaded QA model to find answers for a list of queries in each query's supplied list of Document.
 
-Returns list of dictionaries containing answers sorted by (desc.) probability
+Returns list of dictionaries containing answers sorted by (desc.) score
```
Did you change those manually? The API docs will be generated automatically from the docstrings of the `FARMReader` class (see https://github.com/deepset-ai/haystack/tree/master/docs#update-docstrings)

julian-risch replied:
True. I will revert this change.

julian-risch commented:
> Looks all good to me. Two questions:
> 
> * I assume Summarizer, Re-Ranker and classifier did not use any probabilities before?
> * Should we scale `score` always between 0-1 now? I think this would simplify communication to users a lot. I assume this scaling is currently not the case for Elasticsearch (see comment).

Thanks for the feedback! Summarizer, Re-Ranker and Classifier don't have a `probability` field in the returned Document, correct.

The requested change in the elasticsearch document store is included now and I reverted the manual changes to the `reader.md` file. Further, I implemented the normalization of weights in the JoinDocuments node (as discussed).
